File: js/calc.js
/* ══════════════════════════════════════════════
   calc.js — semua perhitungan saldo

   Sumber kebenaran tunggal adalah MATRIKS akun × kantong.
   Saldo akun    = jumlah baris matriks
   Saldo kantong = jumlah kolom matriks
   Keduanya tidak pernah dihitung terpisah, supaya
   tidak mungkin saling bertentangan.
   ══════════════════════════════════════════════ */

const Calc = {

  /* matriks[akun_id][kantong_id] = nominal */
  matriks(db, sampai) {
    const m = {};
    const batas = sampai ? new Date(sampai).getTime() : Infinity;

    const sel = (ak, kt, delta) => {
      if (!ak || !kt) return;
      if (!m[ak]) m[ak] = {};
      m[ak][kt] = (m[ak][kt] || 0) + delta;
    };

    for (const t of db.transaksi) {
      if (new Date(t.timestamp).getTime() > batas) continue;
      const n = Number(t.nominal) || 0;

      switch (t.jenis) {
        case 'masuk':
        case 'saldo_awal':
          sel(t.akun_id, t.kantong_id, n);
          break;

        case 'keluar':
          sel(t.akun_id, t.kantong_id, -n);
          break;

        /* pindah tempat: kepemilikan tidak berubah */
        case 'transfer_akun':
          sel(t.akun_id,        t.kantong_id, -n);
          sel(t.akun_tujuan_id, t.kantong_id,  n);
          break;

        /* pindah kepemilikan: lokasi fisik tidak berubah */
        case 'transfer_kantong':
          sel(t.akun_id, t.kantong_id,         -n);
          sel(t.akun_id, t.kantong_tujuan_id,   n);
          break;
      }
    }
    return m;
  },

  saldoAkun(db, m) {
    m = m || this.matriks(db);
    const out = {};
    for (const a of db.akun) {
      out[a.id] = Object.values(m[a.id] || {}).reduce((s, v) => s + v, 0);
    }
    return out;
  },

  saldoKantong(db, m) {
    m = m || this.matriks(db);
    const out = {};
    for (const k of db.kantong) out[k.id] = 0;
    for (const ak in m) {
      for (const kt in m[ak]) {
        if (out[kt] === undefined) out[kt] = 0;
        out[kt] += m[ak][kt];
      }
    }
    return out;
  },

  /* Angka-angka utama dashboard */
  ringkas(db) {
    const m  = this.matriks(db);
    const sk = this.saldoKantong(db, m);

    let fisik = 0, milik = 0, titipan = 0;
    const negatif = [];

    for (const k of db.kantong) {
      const v = sk[k.id] || 0;
      fisik += v;
      if (k.jenis === 'titipan') titipan += v; else milik += v;
      if (v < 0) negatif.push({ kantong: k, nilai: v });
    }

    return {
      totalFisik: fisik,
      uangSaya: milik,
      titipan: titipan,
      negatif,                       // sumber dana yang jebol
      saldoAkun: this.saldoAkun(db, m),
      saldoKantong: sk,
      matriks: m
    };
  },

  /* Cek sebelum menyimpan: apakah transaksi ini bikin
     saldo sel matriks jadi minus? (konsep.md §3.2 aturan 1)
     Dikembalikan sebagai peringatan, bukan larangan —
     kenyataan kadang memang begitu, dan menolak input
     hanya membuat orang berhenti mencatat. */
  cekDampak(db, calon) {
    const m = this.matriks(db);
    const cell = (ak, kt) => (m[ak] && m[ak][kt]) || 0;
    const n = Number(calon.nominal) || 0;
    const out = [];

    const periksa = (ak, kt, delta) => {
      if (!ak || !kt) return;
      const sesudah = cell(ak, kt) + delta;
      if (sesudah < 0) {
        out.push({
          akun: db.akun.find(a => a.id === ak),
          kantong: db.kantong.find(k => k.id === kt),
          sesudah
        });
      }
    };

    if (calon.jenis === 'keluar')                periksa(calon.akun_id, calon.kantong_id, -n);
    else if (calon.jenis === 'transfer_akun')    periksa(calon.akun_id, calon.kantong_id, -n);
    else if (calon.jenis === 'transfer_kantong') periksa(calon.akun_id, calon.kantong_id, -n);

    return out;
  },

  /* ID transaksi yang harus dikecualikan dari laporan: transaksi yang
     sudah dikoreksi, beserta entri pembaliknya. Keduanya saling meniadakan.

     Untuk SALDO keduanya justru harus tetap dihitung — hasilnya nol dan
     itu benar. Tapi untuk laporan kategori & arus kas, memasukkan keduanya
     membuat pengeluaran yang sudah dibatalkan tetap muncul di grafik dan
     pembaliknya terhitung sebagai pemasukan. Dua kali salah. */
  idDikoreksi(db) {
    const skip = new Set();
    for (const t of db.transaksi) {
      if (t.reversal_dari) { skip.add(t.id); skip.add(t.reversal_dari); }
    }
    return skip;
  },

  /* Pengeluaran per kategori dalam rentang tanggal.
     transfer sengaja tidak dihitung — itu bukan pengeluaran. */
  perKategori(db, dari, sampai, tipe) {
    tipe = tipe || 'keluar';
    const d = new Date(dari).getTime(), s = new Date(sampai).getTime();
    const skip = this.idDikoreksi(db);
    const agg = {};
    let total = 0;

    for (const t of db.transaksi) {
      if (t.jenis !== tipe) continue;
      if (skip.has(t.id)) continue;
      const w = new Date(t.timestamp).getTime();
      if (w < d || w > s) continue;
      const id = t.kategori_id || '_';
      agg[id] = (agg[id] || 0) + (Number(t.nominal) || 0);
      total += Number(t.nominal) || 0;
    }

    const rows = Object.keys(agg).map(id => {
      const k = db.kategori.find(x => x.id === id);
      return { id, nama: k ? k.nama : 'Tanpa kategori', nilai: agg[id] };
    }).sort((a, b) => b.nilai - a.nilai);

    return { rows, total };
  },

  /* Arus kas periode — hanya dana milik sendiri kalau diminta */
  arusKas(db, dari, sampai, hanyaMilikSendiri) {
    const d = new Date(dari).getTime(), s = new Date(sampai).getTime();
    const skip = this.idDikoreksi(db);
    let masuk = 0, keluar = 0;

    for (const t of db.transaksi) {
      const w = new Date(t.timestamp).getTime();
      if (w < d || w > s) continue;
      if (t.jenis !== 'masuk' && t.jenis !== 'keluar') continue;
      if (skip.has(t.id)) continue;

      if (hanyaMilikSendiri) {
        const k = db.kantong.find(x => x.id === t.kantong_id);
        if (!k || k.jenis !== 'milik_sendiri') continue;
      }
      const n = Number(t.nominal) || 0;
      if (t.jenis === 'masuk') masuk += n; else keluar += n;
    }
    return { masuk, keluar, selisih: masuk - keluar };
  },

  /* Mutasi satu sumber dana — bahan laporan pertanggungjawaban.
     (konsep.md §7) */
  mutasiKantong(db, kantongId, dari, sampai) {
    const d = dari ? new Date(dari).getTime() : -Infinity;
    const s = sampai ? new Date(sampai).getTime() : Infinity;
    const rows = [];
    let saldo = 0;

    const urut = db.transaksi
      .filter(t => t.kantong_id === kantongId || t.kantong_tujuan_id === kantongId)
      .sort((a, b) => new Date(a.timestamp) - new Date(b.timestamp));

    for (const t of urut) {
      const n = Number(t.nominal) || 0;
      let delta = 0;

      if (t.kantong_id === kantongId) {
        if (t.jenis === 'masuk' || t.jenis === 'saldo_awal') delta = n;
        else if (t.jenis === 'keluar') delta = -n;
        else if (t.jenis === 'transfer_kantong') delta = -n;
        // transfer_akun tidak mengubah saldo kantong
      }
      if (t.kantong_tujuan_id === kantongId && t.jenis === 'transfer_kantong') delta = n;

      saldo += delta;
      const w = new Date(t.timestamp).getTime();
      if (w >= d && w <= s && delta !== 0) {
        rows.push({ tx: t, delta, saldo });
      }
    }
    return { rows, saldoAkhir: saldo };
  }
};

File: native/app/src/main/java/id/my/ibstudio/dompetq/domain/Calc.kt
package id.my.ibstudio.dompetq.domain

import id.my.ibstudio.dompetq.data.JENIS_KELUAR
import id.my.ibstudio.dompetq.data.JENIS_MASUK
import id.my.ibstudio.dompetq.data.JENIS_SALDO_AWAL
import id.my.ibstudio.dompetq.data.JENIS_TRANSFER_AKUN
import id.my.ibstudio.dompetq.data.JENIS_TRANSFER_KANTONG
import id.my.ibstudio.dompetq.data.KANTONG_MILIK_SENDIRI
import id.my.ibstudio.dompetq.data.KANTONG_TITIPAN
import id.my.ibstudio.dompetq.data.Akun
import id.my.ibstudio.dompetq.data.Kantong
import id.my.ibstudio.dompetq.data.Kategori
import id.my.ibstudio.dompetq.data.Transaksi

/* ══════════════════════════════════════════════
   Calc — semua perhitungan saldo.

   Sumber kebenaran tunggal adalah MATRIKS akun × kantong.
     saldo akun    = jumlah satu baris
     saldo kantong = jumlah satu kolom
   Keduanya tidak pernah dihitung terpisah, supaya tidak
   mungkin saling bertentangan. Porting langsung dari
   js/calc.js — aturannya harus identik di dua versi.
   ══════════════════════════════════════════════ */

/* matriks[akun_id][kantong_id] = nominal */
typealias Matriks = MutableMap<String, MutableMap<String, Long>>

data class SelMinus(val akun: Akun?, val kantong: Kantong?, val sesudah: Long)

data class BarisKategori(val id: String, val nama: String, val nilai: Long)

data class HasilKategori(val rows: List<BarisKategori>, val total: Long)

data class ArusKas(val masuk: Long, val keluar: Long) {
    val selisih: Long get() = masuk - keluar
}

data class BarisMutasi(val tx: Transaksi, val delta: Long, val saldo: Long)

data class HasilMutasi(val rows: List<BarisMutasi>, val saldoAkhir: Long)

data class Ringkas(
    val totalFisik: Long,
    val uangSaya: Long,
    val titipan: Long,
    val negatif: List<Pair<Kantong, Long>>,
    val saldoAkun: Map<String, Long>,
    val saldoKantong: Map<String, Long>,
    val matriks: Matriks
)

object Calc {

    fun matriks(transaksi: List<Transaksi>, sampai: Long? = null): Matriks {
        val m: Matriks = mutableMapOf()
        val batas = sampai ?: Long.MAX_VALUE

        fun sel(ak: String, kt: String, delta: Long) {
            if (ak.isEmpty() || kt.isEmpty()) return
            val baris = m.getOrPut(ak) { mutableMapOf() }
            baris[kt] = (baris[kt] ?: 0L) + delta
        }

        for (t in transaksi) {
            if (t.timestamp > batas) continue
            val n = t.nominal

            when (t.jenis) {
                JENIS_MASUK, JENIS_SALDO_AWAL -> sel(t.akun_id, t.kantong_id, n)
                JENIS_KELUAR -> sel(t.akun_id, t.kantong_id, -n)

                /* pindah tempat: kepemilikan tidak berubah */
                JENIS_TRANSFER_AKUN -> {
                    sel(t.akun_id, t.kantong_id, -n)
                    sel(t.akun_tujuan_id, t.kantong_id, n)
                }

                /* pindah kepemilikan: lokasi fisik tidak berubah */
                JENIS_TRANSFER_KANTONG -> {
                    sel(t.akun_id, t.kantong_id, -n)
                    sel(t.akun_id, t.kantong_tujuan_id, n)
                }
            }
        }
        return m
    }

    fun saldoAkun(akun: List<Akun>, m: Matriks): Map<String, Long> =
        akun.associate { a -> a.id to (m[a.id]?.values?.sum() ?: 0L) }

    fun saldoKantong(kantong: List<Kantong>, m: Matriks): Map<String, Long> {
        val out = kantong.associate { it.id to 0L }.toMutableMap()
        for ((_, kolom) in m) {
            for ((kt, v) in kolom) out[kt] = (out[kt] ?: 0L) + v
        }
        return out
    }

    fun ringkas(akun: List<Akun>, kantong: List<Kantong>, transaksi: List<Transaksi>): Ringkas {
        val m = matriks(transaksi)
        val sk = saldoKantong(kantong, m)

        var fisik = 0L
        var milik = 0L
        var titipan = 0L
        val negatif = mutableListOf<Pair<Kantong, Long>>()

        for (k in kantong) {
            val v = sk[k.id] ?: 0L
            fisik += v
            if (k.jenis == KANTONG_TITIPAN) titipan += v else milik += v
            if (v < 0) negatif += k to v
        }

        return Ringkas(fisik, milik, titipan, negatif, saldoAkun(akun, m), sk, m)
    }

    /* Cek sebelum menyimpan: apakah transaksi ini bikin satu sel matriks
       jadi minus? Dikembalikan sebagai PERINGATAN, bukan larangan —
       kenyataan kadang memang begitu, dan menolak input hanya membuat
       orang berhenti mencatat. (konsep.md §3.2 aturan 1) */
    fun cekDampak(
        akun: List<Akun>,
        kantong: List<Kantong>,
        transaksi: List<Transaksi>,
        calon: Transaksi
    ): List<SelMinus> {
        val m = matriks(transaksi)
        val out = mutableListOf<SelMinus>()

        fun periksa(ak: String, kt: String, delta: Long) {
            if (ak.isEmpty() || kt.isEmpty()) return
            val sesudah = (m[ak]?.get(kt) ?: 0L) + delta
            if (sesudah < 0) {
                out += SelMinus(
                    akun.firstOrNull { it.id == ak },
                    kantong.firstOrNull { it.id == kt },
                    sesudah
                )
            }
        }

        when (calon.jenis) {
            JENIS_KELUAR, JENIS_TRANSFER_AKUN, JENIS_TRANSFER_KANTONG ->
                periksa(calon.akun_id, calon.kantong_id, -calon.nominal)
        }
        return out
    }

    /* ID transaksi yang harus dikecualikan dari LAPORAN: yang sudah
       dikoreksi beserta entri pembaliknya.

       Untuk SALDO keduanya justru tetap dihitung — hasilnya nol dan itu
       benar. Tapi di laporan kategori, memasukkan keduanya membuat
       pengeluaran yang sudah dibatalkan tetap muncul di grafik dan
       pembaliknya terhitung sebagai pemasukan. Dua kali salah. */
    fun idDikoreksi(transaksi: List<Transaksi>): Set<String> {
        val skip = mutableSetOf<String>()
        for (t in transaksi) {
            if (t.reversal_dari.isNotEmpty()) {
                skip += t.id
                skip += t.reversal_dari
            }
        }
        return skip
    }

    /* Pengeluaran per kategori dalam rentang tanggal.
       transfer sengaja tidak dihitung — itu bukan pengeluaran. */
    fun perKategori(
        transaksi: List<Transaksi>,
        kategori: List<Kategori>,
        dari: Long,
        sampai: Long,
        tipe: String = JENIS_KELUAR
    ): HasilKategori {
        val skip = idDikoreksi(transaksi)
        val agg = mutableMapOf<String, Long>()
        var total = 0L

        for (t in transaksi) {
            if (t.jenis != tipe) continue
            if (t.id in skip) continue
            if (t.timestamp < dari || t.timestamp > sampai) continue
            val id = t.kategori_id.ifEmpty { "_" }
            agg[id] = (agg[id] ?: 0L) + t.nominal
            total += t.nominal
        }

        val rows = agg.map { (id, nilai) ->
            BarisKategori(id, kategori.firstOrNull { it.id == id }?.nama ?: "Tanpa kategori", nilai)
        }.sortedByDescending { it.nilai }

        return HasilKategori(rows, total)
    }

    /* Arus kas periode — bisa dibatasi hanya dana milik sendiri, supaya
       uang titipan yang lewat tidak terbaca sebagai pemasukan pribadi. */
    fun arusKas(
        transaksi: List<Transaksi>,
        kantong: List<Kantong>,
        dari: Long,
        sampai: Long,
        hanyaMilikSendiri: Boolean
    ): ArusKas {
        val skip = idDikoreksi(transaksi)
        var masuk = 0L
        var keluar = 0L

        for (t in transaksi) {
            if (t.timestamp < dari || t.timestamp > sampai) continue
            if (t.jenis != JENIS_MASUK && t.jenis != JENIS_KELUAR) continue
            if (t.id in skip) continue

            if (hanyaMilikSendiri) {
                val k = kantong.firstOrNull { it.id == t.kantong_id }
                if (k == null || k.jenis != KANTONG_MILIK_SENDIRI) continue
            }
            if (t.jenis == JENIS_MASUK) masuk += t.nominal else keluar += t.nominal
        }
        return ArusKas(masuk, keluar)
    }

    /* Mutasi satu sumber dana — bahan laporan pertanggungjawaban.
       (konsep.md §7) */
    fun mutasiKantong(
        transaksi: List<Transaksi>,
        kantongId: String,
        dari: Long? = null,
        sampai: Long? = null
    ): HasilMutasi {
        val d = dari ?: Long.MIN_VALUE
        val s = sampai ?: Long.MAX_VALUE
        val rows = mutableListOf<BarisMutasi>()
        var saldo = 0L

        val urut = transaksi
            .filter { it.kantong_id == kantongId || it.kantong_tujuan_id == kantongId }
            .sortedBy { it.timestamp }

        for (t in urut) {
            var delta = 0L

            if (t.kantong_id == kantongId) {
                delta = when (t.jenis) {
                    JENIS_MASUK, JENIS_SALDO_AWAL -> t.nominal
                    JENIS_KELUAR -> -t.nominal
                    JENIS_TRANSFER_KANTONG -> -t.nominal
                    /* transfer_akun tidak mengubah saldo kantong */
                    else -> 0L
                }
            }
            if (t.kantong_tujuan_id == kantongId && t.jenis == JENIS_TRANSFER_KANTONG) {
                delta = t.nominal
            }

            saldo += delta
            if (t.timestamp in d..s && delta != 0L) rows += BarisMutasi(t, delta, saldo)
        }
        return HasilMutasi(rows, saldo)
    }
}
